Add batch product update endpoint

wzieba opened the pull request:
Closes: https://github.com/woocommerce/woocommerce-android/issues/7931

For testing instructions, please see Woo PR: https://github.com/woocommerce/woocommerce-android/pull/7965

wzieba committed:
feat: add batch product update endpoint

wzieba committed:
feat: add store method to update list of products

wzieba committed:
feat: use database product models in payload

wzieba committed:
style: fix empty line

wzieba committed:
Merge remote-tracking branch 'origin/trunk' into woo/7931_batch_update_products

wzieba committed:
Merge remote-tracking branch 'origin/trunk' into woo/7931_batch_update_products

atorresveiga commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductApiResponse.kt:
```diff
@@ -4,89 +4,76 @@ import com.google.gson.JsonArray
+typealias ProductDto = ProductApiResponse
+
+@Suppress("ConstructorParameterNaming")
+data class ProductApiResponse(
```
Nice!

atorresveiga commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt:
```diff
@@ -931,6 +931,30 @@ class ProductRestClient @Inject constructor(
+    ): WooPayload<BatchProductApiResponse> = WOOCOMMERCE.products.batch.pathV3
+        .let { url ->
+            val body = buildMap {
+                putIfNotNull("update" to (existingProducts zip updatedProducts).map { (existing, updated) ->
```
The `zip` operation will return a list of pairs taking into account the indexes of the two collections. I believe we must sort these two collections before applying the zip to ensure that the products have the same indexes and that, with the `zip` operation, the pair will be formed with the `same` product.

In the example above, the item at index `0` is different in both collections (updateProducts vs existingProducts)

<img width="500" alt="Screen Shot 2022-12-01 at 08 39 59" src="https://user-images.githubusercontent.com/18119390/205084047-b3566680-222c-4430-a647-345ee1235d69.png">

<img width="300" alt="Screen Shot 2022-12-01 at 08 41 34" src="https://user-images.githubusercontent.com/18119390/205084056-7a83435e-91b0-4b27-beaa-7915f2f3b0d9.png">

wzieba replied:
Thanks @atorresveiga , good point! Fixed in bb41344e2ff855061c8f8ddeb10160465c1501f8

atorresveiga replied:
Wow! That was fast 🚀. Thanks

wzieba committed:
fix: assert that existing/updated product pair match

atorresveiga commented:
Great work @wzieba! 
Kudos for refactoring the `ProductApiResponse` and converting it to a data class, and using `productModelToProductJsonBody` to send only the necessary field to update.
I left one comment about sorting the list to ensure `updateProducts` and `existingProducts` have the same order. Maybe we can unit-test this and add other tests to the `batchUpdateProducts`.🤔
It would be nice to add this batch update function to the example app on FluxC, so we can test the endpoint in isolation without needing the Woo app. Something that is not required for this PR, but a nice to have for the future.

atorresveiga commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt:
```diff
@@ -931,6 +931,33 @@ class ProductRestClient @Inject constructor(
+                putIfNotNull("update" to sortedExistingToUpdatedProducts.map { (existing, updated) ->
+                        productModelToProductJsonBody(
+                            productModel = existing, updatedProductModel = updated
+                        ).plus("id" to updated.remoteProductId)
```
np: Adding the `plus` without checking if the map is empty will always send the product, even if it doesn't have any change. WDYT, if we check the map size first and send only products with changes?

<img width="359" alt="Screen Shot 2022-12-01 at 10 21 44" src="https://user-images.githubusercontent.com/18119390/205105994-5f3260e8-2e75-46e1-a716-39432138f883.png">

wzieba committed:
tests: add unit tests for sorting behaviour

wzieba committed:
fix: not send update request with id if no changes

wzieba commented:
> I left one comment about sorting the list to ensure updateProducts and existingProducts have the same order. Maybe we can unit-test this and add other tests to the batchUpdateProducts.🤔

Thanks. I've moved sorting from `RestClient` to `Store` (as I think it's rather `Store` responsibility to prepare correct data for `RestClient` to operate) and that allowed me to add a unit test.

I've also added your second improvement: 
>check the map size first and send only products with changes

Ready for the second round!

atorresveiga commented:
@wzieba I tested the code and found this:

<img width="378" alt="Screen Shot 2022-12-02 at 13 16 03" src="https://user-images.githubusercontent.com/18119390/205377860-1a464b8f-155b-4bf4-9fdf-fe6a2aee6520.png">

<img width="524" alt="Screen Shot 2022-12-02 at 14 03 53" src="https://user-images.githubusercontent.com/18119390/205377863-c2a2da7e-26eb-400a-b8fc-6ba41181a1d5.png">

Note that in this example, the `id` property is not added to the HashMap.

Maybe something like this could do the trick:
```
putIfNotNull("update" to existingToUpdatedProducts.mapNotNull { (existing, updated) ->
    val jsonBody = productModelToProductJsonBody(
        productModel = existing, updatedProductModel = updated
    )
    if (jsonBody.isNotEmpty()) {
        jsonBody.plus("id" to updated.remoteProductId)
    }else{
        null
    }
}
```

wzieba committed:
Merge remote-tracking branch 'origin/trunk' into woo/7931_batch_update_products

wzieba committed:
tests: add unit tests for preparing body in rest client

wzieba committed:
fix: do not send any payload if products don't differ

wzieba commented:
Thanks, @atorresveiga , you're entirely right. I've rushed with this change.

In 36da5cc I've added unit tests and then adjusted behavior in c575c8d. It should work fine now.

atorresveiga commented:
Great work @wzieba! Nice touch of adding unit tests to the request generation. I tested the code, and it is working great. I've one last nitpick advice, though. Maybe we can use `existingToUpdatedProducts.mapNotNull` instead of `existingToUpdatedProducts.map`. This way, only values that are different than `null` will be added to the list of products to update. 
Using `map` could end with a request/response like the one below, sending a `null` value that is not required and, because of that, receiving an error response between the results:
<img width="478" alt="Screen Shot 2022-12-05 at 09 51 43" src="https://user-images.githubusercontent.com/18119390/205681839-8fbfcdb0-0a08-489b-9a30-8c6cb587104e.png">

wzieba committed:
fix: do not add null values to payload if no changes

wzieba commented:
Thank you again @atorresveiga , fix applied in 070238ffedee775ef8e8f1c50d8f45ae4cb651b3

atorresveiga commented:
Great work, @wzieba! 
LGTM! :shipit:

wzieba commented:
You have to approve the PR @atorresveiga to make all check green 😊

atorresveiga reviewed (approved):
LGTM!
